wallet/walletrpc.c

```diff
@@ -177,29 +177,54 @@ static const struct json_command newaddr_command = {
 };
 AUTODATA(json_command, &newaddr_command);
 
-static struct command_result *json_listaddrs(struct command *cmd,
+static void json_add_addr(struct json_stream *response,
+				 const u64 keyidx,
+				 const struct pubkey *pubkey,
+				 const char *out_p2wpkh,
+				 const u8 *redeemscript_p2wpkh,
+				 const char *out_p2tr)
+{
+	json_object_start(response, NULL);
+	json_add_u64(response, "keyidx", keyidx);
+	json_add_pubkey(response, "pubkey", pubkey);
+	json_add_string(response, "bech32", out_p2wpkh);
+	json_add_hex_talarr(response, "bech32_redeemscript", redeemscript_p2wpkh);
+	json_add_string(response, "p2tr", out_p2tr);
+	json_object_end(response);
+}
+
+static struct command_result *json_listaddresses(struct command *cmd,
 					     const char *buffer,
 					     const jsmntok_t *obj UNNEEDED,
 					     const jsmntok_t *params)
 {
 	struct json_stream *response;
 	struct pubkey pubkey;
-	u64 *bip32_max_index;
+	const u8 *address;
+	u64 *start, *end;
+	bool *allow_future_addr;
 
 	if (!param(cmd, buffer, params,
-		   p_opt("bip32_max_index", param_u64, &bip32_max_index),
-		   NULL))
+			 p_opt_def("start", param_u64, &start, 0),
+			 p_opt("end", param_u64, &end),
+			 p_opt("address", param_bitcoin_address, &address),
+			 p_opt_def("allow_future_addr", param_bool, &allow_future_addr, false),
+			 NULL))
 		return command_param_failed();
 
-	if (!bip32_max_index) {
-		bip32_max_index = tal(cmd, u64);
-		*bip32_max_index = db_get_intvar(cmd->ld->wallet->db,
-						 "bip32_max_index", 0);
+	u64 *bip32_max_index = tal(cmd, u64);
+	*bip32_max_index = db_get_intvar(cmd->ld->wallet->db, "bip32_max_index", 0);
+	if (!end || (!*allow_future_addr && *end > *bip32_max_index)) {
+		end = bip32_max_index;
 	}
+
+	if (command_check_only(cmd))
+		return command_check_done(cmd);
+
```

Comment:
You don't need this, since you used "param" (which does this internally) not "param_check"

Reply:
Removed, thanks!